### ISSUE TITLE
Support Unicode strings in CHAPI

### ISSUE THREAD

**Issue by MarshallBrinn** [GENI-NSF/geni-ch#310]
CHAPI needs to support reading and writing unicode (non-ascii) strings in its API's and database interactions.

 _Imported from trac ticket [#310](http://trac.gpolab.bbn.com/chapi/ticket/310),  created by mbrinn on 06-12-2014 at 09:32, last modified: 08-13-2014 at 15:03_

**Comment by MarshallBrinn** [comment]
This ticket is a partner of PROTO-CH ticket 1069 http://trac.gpolab.bbn.com/proto-ch/ticket/1069

_Trac comment by mbrinn (github user: MarshallBrinn) on 06-12-2014 at 09:34_

**Comment by MarshallBrinn** [comment]
Summarizing some investigations about Unicode and UTF-8.

First, for clarity: Unicode is a type of string that contains international characters. It can be encoded in a number of ways into byte strings, the most popular of which is UTF-8. That is, you ENCODE Unicode into UTF-8, and you DECODE UTF-8 into Unicode. UTF-8 encodings are strings not unicode:

To illustrate, here's some python:
```
>>> name = 'Ren\xc3\xa9'
>>> uname = name.decode('utf-8')
>>> name
'Ren\xc3\xa9'
>>> uname
u'Ren\xe9'
>>> 
```

The variable name is a utf-8 encoding of the name René. The variable uname is a unicode value of the name René (note it has a u'' prefix.

We want our databases to use a UTF8 encoding, rather than the default SQL_ASCII encoding. In this encoding, it takes utf-8 strings as well as unicode strings and stores them as UTF-8, and then retrieves them as unicode.

Here's how to create a database with UTF8 encoding.

```
sudo -u postgres /usr/bin/createdb -E UTF8 -T template0 test_utf8
```

Within that database, all tables will be created with character_encoding UTF8.

Our plan for the CHAPI and Portal databases is as follows:

1. Dump the portal database
2. Drop the portal database
3. Create a new database with name portal and UTF8 encoding.
4. Load the database from the dump file.

The general principle should be that the PHP and Python processes work with Unicode, and only change to UTF-8 at the edges. The database should do this transparently (no need for us to decode/encode).

Then we may have to encode unicode strings for displaying in HTML or emails (the other user facing edges).

A few more details if you want to mess with this.

To tell the encoding of databases, use the \l command in psql.

To access the database in sqlalchemy/python, use:

```
import sqlalchemy
# Change the URL as needed which is user:pwd@host/dbname
engine = sqlalchemy.create_engine('postgresql://portal:portal@localhost/portal')
md = sqlalchemy.MetaData(engine)
# Map whatever tables you want
test_table = sqlalchemy.Table('test', md, autoload=True)
conn = engine.connect()
# Query from database, change query as needed
res = conn.execute("Select * from test")
# Insert into database. Uname in this case is a unicode string to a varchar field, which will be converted to/from UTF-8
ins = test_table.insert()values(id=6, value=uname, description='SQLAlchemy direct unicode insert')
conn.execute(ins)

```


_Trac comment by mbrinn (github user: MarshallBrinn) on 06-12-2014 at 09:50_

**Comment by MarshallBrinn** [comment]
Here's a quote from http://blog.notdot.net/2010/07/Getting-unicode-right-in-Python that gives a good definition of how an application should be handling unicode and utf-8:

The correct solution is to fix your code so that you handle text correctly. Here's the cliff's notes on what you should be doing:
 * All text strings, everywhere should be of type unicode, not str. If you're handling text, and your variable is a str, it's a bug!
 * To decode a byte string as text, use var.decode(encoding) (eg, var.decode('utf-8'), with the correct encoding. To encode a text string as bytes, use var.encode(encoding).
 * Never ever use str() on a unicode string, or unicode() on a byte string without a second argument specifying the encoding.  
 * Whenever you read data from outside your app, expect it to be bytes - eg, of type str - and call .decode() on it to interpret it as text. Likewise, always call .encode() on text you want to send to the outside world.
 * If a string literal in your code is intended to represent text, it should always be prefixed with 'u'. In fact, you probably never want to define a raw string literal in your code at all. For what it's worth, though, I'm terrible at this one, as I'm sure pretty much everyone else is, too.


_Trac comment by mbrinn (github user: MarshallBrinn) on 06-12-2014 at 15:20_

**Comment by MarshallBrinn** [comment]
This seems to be all working on the i18n branch pending inspection/testing.

The main things to note:
 * Need to drop and recreate the portal database with UTF8 encoding
 * Everything should be unicode in both portal and chapi code except at the edges.
 * The strings that come in from the IDP are UTF8 encoded. We probably should treat this as an 'edge' and translate as they come in.
 * The strings need to be UTF8 encoded when writing emails.
 * The strings need to be UTF8 encoded when writing directly to the database from the PHP code. Python SQL alchemy handles the Unicode/UTF8 translation seamlessly. But PHP just wants a simple 'insert' statement and that needs to be in UTF-8. Look at the identity_attribute table.


_Trac comment by mbrinn (github user: MarshallBrinn) on 06-30-2014 at 10:42_

**Comment by MarshallBrinn** [comment]
Here are the instructions for transforming the portal database from SQL_ASCII to UTF8:

Change portal database to UTF8:

# Make copy of database
pg_dump -U portal -h localhost portal > /tmp/portal.sql

# See the original database as SQL_ASCII
sudo -u postgres psql -l
                              List of databases
   Name    |  Owner   | Encoding  | Collation | Ctype |   Access privileges
-----------+----------+-----------+-----------+-------+-----------------------
 portal    | postgres | SQL_ASCII | C         | C     |
 postgres  | postgres | SQL_ASCII | C         | C     |
 template0 | postgres | SQL_ASCII | C         | C     | =c/postgres
                                                      : postgres=CTc/postgres
 template1 | postgres | SQL_ASCII | C         | C     | =c/postgres
                                                      : postgres=CTc/postgres
 test_utf8 | postgres | UTF8      | C         | C     |



# Stop apache
sudo service apache2 stop
 * Stopping web server apache2
 ... waiting    ...done.

# Drop the portal database
sudo -u postgres dropdb portal

# Create the portal databse
sudo -u postgres createdb -E UTF8 -T template0 portal

# Verify the database
sudo -u postgres psql -l
                              List of databases
   Name    |  Owner   | Encoding  | Collation | Ctype |   Access privileges
-----------+----------+-----------+-----------+-------+-----------------------
 portal    | postgres | UTF8      | C         | C     |
 postgres  | postgres | SQL_ASCII | C         | C     |
 template0 | postgres | SQL_ASCII | C         | C     | =c/postgres
                                                      : postgres=CTc/postgres
 template1 | postgres | SQL_ASCII | C         | C     | =c/postgres
                                                      : postgres=CTc/postgres
 test_utf8 | postgres | UTF8      | C         | C     |
(5 rows)

psql -U portal -h localhost portal < /tmp/portal.sql



_Trac comment by mbrinn (github user: MarshallBrinn) on 07-01-2014 at 09:27_

**Comment by MarshallBrinn** [comment]
Closed along with partner ticket #1069.

_Trac comment by mbrinn (github user: MarshallBrinn) on 08-13-2014 at 15:03_
